src/main/java/org/truffleruby/core/array/ArrayNodes.java

```diff
@@ -781,6 +781,21 @@ protected boolean equalSamePrimitiveType(VirtualFrame frame, DynamicObject a, Dy
             return true;
         }
 
+        @Specialization(guards = { "isRubyArray(b)", "slowStrategy(a).matches(b)" }, replaces = "equalSamePrimitiveType")
+        protected boolean equalSamePrimitiveTypeGeneric(VirtualFrame frame, DynamicObject a, DynamicObject b,
+                @Cached("createBinaryProfile()") ConditionProfile sameProfile,
+                @Cached("createIdentityProfile()") IntValueProfile sizeProfile,
+                @Cached("createBinaryProfile()") ConditionProfile sameSizeProfile,
+                @Cached("create()") BranchProfile trueProfile,
+                @Cached("create()") BranchProfile falseProfile) {
+            return equalSamePrimitiveType(frame, a, b, slowStrategy(a), sameProfile, sizeProfile, sameSizeProfile, trueProfile, falseProfile);
+        }
```

Comment:
We should not need this specialization because ARRAY_STRATEGIES is large enough to accommodate every combination of specializations.
Do you have a case where this node fails unexpectedly and this fixes it?
Or is it for new DSL features/restrictions?

Reply:
The new version of the DSL will spit out the following error:
```Some guards for the following specializations could not be negated for the @Fallback specialization: [equalSamePrimitiveType]. Guards cannot be negated for the @Fallback when they bind @Cached parameters and the specialization may consist of multiple instances. To fix this limit the number of instances to '1' or introduce a more generic specialization declared between this specialization and the fallback. Alternatively the use of @Fallback can be avoided by declaring a @Specialization with manually specified negated guards.
```

Reply:
In other words the current node can crash if you trigger the fallback first.

Reply:
fallbacks for inline caches with constant limits are pretty hard to support, fallbacks for inline caches without constant limits are basically even harder as they always need to invoke the limit check as guard.

Reply:
I see, supporting `@Fallback` seems fairly complicated.
In this specific case, it seems cleaner to just have a guard `!isRubyArray(b)` for the fallback.
I'll make a commit with that.